### PR TITLE
feat:[CI-15370]: Kaniko push only support to push local tarball image.

### PR DESCRIPTION
**Pull request by Anshika2203**
Used crane to push images to registry.

Sample execution using a large image (pytorch)
https://app.harness.io/ng/account/gCoPSwHxS7ipOgx2iA9tOQ/all/orgs/default/projects/anshikaci/pipelines/CI15370_CheckingLocalTarballPushKaniko/deployments/bd-HHwVRQf6rMR8Gbt5qFQ/pipeline?storeType=INLINE

### CODE REVIEW

File: kaniko.go [hunk 1]
```diff
@@ -174,6 +176,50 @@ func (p Plugin) Exec() error {
...
+		for _, tag := range tags {
+			dest := fmt.Sprintf("%s:%s", p.Build.Repo, tag)
+
+			// Push the image to the destination
+			err := crane.Push(img, dest)
+			if err != nil {
+				return fmt.Errorf("failed to push image to %s: %v", dest, err)
+			}
+
+			fmt.Printf("Successfully pushed %s\n", dest)
```
Comment (hemanthmantri) [inline kaniko.go:217]: Let's include image name and destination in the log message.

File: kaniko.go [hunk 2]
```diff
@@ -174,6 +176,50 @@ func (p Plugin) Exec() error {
...
+				return err
+			}
+		} else if len(p.Build.Tags) > 0 {
+			tags = p.Build.Tags
+		} else {
+			tags = []string{"latest"}
+		}
+
+		// Load the image from the tarball
+		img, err := crane.Load(p.Build.LocalTarPath)
```
Comment (hemanthmantri) [inline kaniko.go:203]: Are there any size limits for the image? Please test with large images.

File: kaniko.go [hunk 3]
```diff
@@ -174,6 +176,50 @@ func (p Plugin) Exec() error {
...
+		if p.Build.AutoTag {
+			tags, err = p.Build.AutoTags()
+			if err != nil {
+				return err
+			}
+		} else if len(p.Build.Tags) > 0 {
+			tags = p.Build.Tags
+		} else {
+			tags = []string{"latest"}
+		}
```
Comment (hemanthmantri) [inline kaniko.go:200]: Is this current behavior when no tags are specified?

File: cmd/kaniko-docker/main.go [hunk 4]
```diff
@@ -448,6 +453,7 @@ func run(c *cli.Context) error {
 			SkipTLSVerifyRegistry:       c.Bool("skip-tls-verify-registry"),
 			UseNewRun:                   c.Bool("use-new-run"),
 			IgnorePath:                  c.String("ignore-path"),
+			LocalTarPath:                c.String("local-tar-path"),
```
Comment (hemanthmantri) [inline cmd/kaniko-docker/main.go:456]: Place options/fields in an alphabetical order.

Comment (hemanthmantri) [reply cmd/kaniko-docker/main.go:456]: Nit: `ImageTarPath` ?

File: kaniko.go (hunk 2 above)
Comment (Anshika2203) [reply kaniko.go:203]: Yep doing that only right now.

File: kaniko.go (hunk 1 above)
Comment (Anshika2203) [reply kaniko.go]: Included. ✅

File: cmd/kaniko-docker/main.go (hunk 4 above)
Comment (Anshika2203) [reply cmd/kaniko-docker/main.go]: Changed `LocalTarPath` to `ImageTarPath` and arranged them in alphabetical order. ✅

File: kaniko.go (hunk 3 above)
Comment (Anshika2203) [reply kaniko.go:200]: Yes, if no tags are mentioned it pushes as `latest` tag.
It's the default behaviour for kaniko/docker.

File: kaniko.go (hunk 2 above)
Comment (Anshika2203) [reply kaniko.go]: I tried with huge pytorch image (~4.5gb just for linux amd64 image) just to check and it pushed successfully to the registry.
execution - https://app.harness.io/ng/account/gCoPSwHxS7ipOgx2iA9tOQ/all/orgs/default/projects/anshikaci/pipelines/CI15370_CheckingLocalTarballPushKaniko/deployments/bd-HHwVRQf6rMR8Gbt5qFQ/pipeline?storeType=INLINE&step=4pm27Kp8SFer4TxPrI8YhA&stage=zmENeIDmQa-c_B5fGwcrXA&childStage=&stageExecId=

File: kaniko.go [hunk 5]
```diff
@@ -174,6 +182,47 @@ func (p Plugin) Exec() error {
...
+		img, err := crane.Load(p.Build.SourceTarPath)
+		if err != nil {
+			return fmt.Errorf("failed to load image from tarball: %v", err)
+		}
+
+		// If no tags are specified, use 'latest'
+		tags := []string{"latest"}
+		if len(p.Build.Tags) > 0 {
+			tags = p.Build.Tags
+		}
```
Comment (Ompragash) [inline kaniko.go:209]:
```suggestion
		tags := p.Build.Tags
```

this is redundant as in main.go it is already ensured that the default value for `tags` is `latest` via the CLI flag

File: kaniko.go [hunk 6]
```diff
@@ -174,6 +182,47 @@ func (p Plugin) Exec() error {
...
+		if p.Build.SourceTarPath == "" {
+			return fmt.Errorf("source_tar_path is required when push_only is set. Please provide a valid tarball path.")
+		}
+
+		if _, err := os.Stat(p.Build.SourceTarPath); os.IsNotExist(err) {
+			return fmt.Errorf("image tarball does not exist at path: %s", p.Build.SourceTarPath)
+		}
+
+		if p.Build.Repo == "" {
+			return fmt.Errorf("destination repository name must be specified when push-only is set")
```
Comment (Ompragash) [inline kaniko.go:196]:
```suggestion
			return fmt.Errorf("missing required destination repository for push-only operation"")
```

File: kaniko.go [hunk 7]
```diff
@@ -174,6 +182,47 @@ func (p Plugin) Exec() error {
...
+			tags = p.Build.Tags
+		}
+
+		for _, tag := range tags {
+			dest := fmt.Sprintf("%s:%s", p.Build.Repo, tag)
+
+			// Push the image to the destination
+			err := crane.Push(img, dest)
+			if err != nil {
+				return fmt.Errorf("failed to push image to %s: %v", dest, err)
```
Comment (Ompragash) [inline kaniko.go:217]:
```suggestion
				return fmt.Errorf("failed to push image from tarball [%s] to destination [%s]: %v", p.Build.SourceTarPath, dest, err)
```

File: kaniko_test.go [hunk 8]
```diff
@@ -280,3 +284,230 @@ func TestTarPathValidation(t *testing.T) {
 		})
 	}
 }
+
+func TestLocalTarballPush(t *testing.T) {
+	tests := []struct {
+		name          string
+		localTarPath  string
```
Comment (Ompragash) [inline kaniko_test.go:291]:
```suggestion
		sourceTarPath  string
```
update all the refs to sourceTarPath

File: kaniko_test.go [hunk 9]
```diff
@@ -280,3 +284,230 @@ func TestTarPathValidation(t *testing.T) {
...
+	return "", nil
+}
+
+func (m *mockImage) ConfigFile() (*v1.ConfigFile, error) {
+	return nil, nil
+}
+
+func (m *mockImage) ConfigName() (v1.Hash, error) {
+	return v1.Hash{}, nil
+}
```
Comment (Ompragash) [inline kaniko_test.go:416]: let's move this to a dedicated test util file `mock_crane.go` and define mock functions `MockCraneLoad` `MockCranePush` for tests - then this mocks can be directly used in kaniko_test.go

File: kaniko.go
Comment (Ompragash) [inline kaniko.go:1]: add the following check in the Exec function at the start as no-push and push-only should be treated as mutually exclusive inputs
```go
	if p.Build.NoPush && p.Build.PushOnly {
		return fmt.Errorf("inputs no-push and push-only cannot be used together. Please define only one")
	}
```

**Review by Ompragash — CHANGES_REQUESTED** [review]
README should also be updated as needed + the below should be documented under `Usage` section

```md
- If `no-push` and `destination_tar_path` are defined: Perform a build only operation and store the resulting image tarball at the specified destination_tar_path.
- If `push-only` and `source_tar_path` are defined: Perform a push only operation by loading the image tarball from the specified `source_tar_path` and pushing it to a Container Registry.
- If both `no-push` and `push-only` inputs are provided, then the plugin throws an error and terminate the operation with a clear message: “Inputs no-push and push-only cannot be used together. Please define only one.”
- If neither `no-push` nor `push-only` is provided: Default to the current plugin behavior (e.g., building and pushing images).
```

File: kaniko.go
Comment (Anshika2203) [reply kaniko.go:1]: Added. ✅

File: kaniko_test.go (hunk 9 above)
Comment (Anshika2203) [reply kaniko_test.go]: Done. ✅

File: kaniko_test.go (hunk 8 above)
Comment (Anshika2203) [reply kaniko_test.go]: Updated. ✅

**Pull request closed without merging**